Fixed einsum compatibility/performance issues

heitorschueroff opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#46398 Fixed einsum compatibility/performance issues**

#### Update: Fixed BC breaking change after reverting

This PR makes torch.einsum compatible with numpy.einsum except for the sublist input option as requested here https://github.com/pytorch/pytorch/issues/21412. It also fixed 2 performance issues linked below and adds a check for reducing to torch.dot instead of torch.bmm which is faster in some cases.

fixes #45854, #37628, #30194, #15671

fixes #41467 with benchmark below
```python
import torch
from torch.utils.benchmark import Timer

a = torch.randn(10000, 100, 101, device='cuda')
b = torch.randn(10000, 101, 3, device='cuda')

c = torch.randn(10000, 100, 1, device='cuda')
d = torch.randn(10000, 100, 1, 3, device='cuda')

print(Timer(
    stmt='torch.einsum("bij,bjf->bif", a, b)',
    globals={'a': a, 'b': b}
).blocked_autorange())

print()

print(Timer(
    stmt='torch.einsum("bic,bicf->bif", c, d)',
    globals={'c': c, 'd': d}
).blocked_autorange())
```
```
<torch.utils.benchmark.utils.common.Measurement object at 0x7fa37c413850>
torch.einsum("bij,bjf->bif", a, b)
  Median: 4.53 ms
  IQR:    0.00 ms (4.53 to 4.53)
  45 measurements, 1 runs per measurement, 1 thread

<torch.utils.benchmark.utils.common.Measurement object at 0x7fa37c413700>
torch.einsum("bic,bicf->bif", c, d)
  Median: 63.86 us
  IQR:    1.52 us (63.22 to 64.73)
  4 measurements, 1000 runs per measurement, 1 thread
```

fixes #32591 with benchmark below
```python
import torch
from torch.utils.benchmark import Timer

a = torch.rand(1, 1, 16, 2, 16, 2, 16, 2, 2, 2, 2, device="cuda")
b = torch.rand(729, 1, 1, 2, 1, 2, 1, 2, 2, 2, 2, device="cuda")

print(Timer(
    stmt='(a * b).sum(dim = (-3, -2, -1))',
    globals={'a': a, 'b': b}
).blocked_autorange())

print()

print(Timer(
    stmt='torch.einsum("...ijk, ...ijk -> ...", a, b)',
    globals={'a': a, 'b': b}
).blocked_autorange())
```
```
<torch.utils.benchmark.utils.common.Measurement object at 0x7efe0de28850>
(a * b).sum(dim = (-3, -2, -1))
  Median: 17.86 ms
  2 measurements, 10 runs per measurement, 1 thread

<torch.utils.benchmark.utils.common.Measurement object at 0x7efe0de286a0>
torch.einsum("...ijk, ...ijk -> ...", a, b)
  Median: 296.11 us
  IQR:    1.38 us (295.42 to 296.81)
  662 measurements, 1 runs per measurement, 1 thread
```

TODO

- [x] add support for ellipsis broadcasting
- [x] fix corner case issues with sumproduct_pair
- [x] update docs and add more comments
- [x] add tests for error cases

Differential Revision: [D24860367](https://our.internmc.facebook.com/intern/diff/D24860367)

heitorschueroff committed:
Fixed einsum compatibility issues

[ghstack-poisoned]

heitorschueroff committed:
Update on "Fixed einsum compatibility issues"


TODO

- [ ] add support for ellipsis broadcasting
- [ ] fix corner case issues with sumproduct_pair
- [ ] update docs and add more comments

[ghstack-poisoned]

heitorschueroff committed:
Update on "Fixed einsum compatibility issues"


TODO

- [x] add support for ellipsis broadcasting
- [ ] fix corner case issues with sumproduct_pair
- [ ] update docs and add more comments
- [ ] add tests for error cases

[ghstack-poisoned]

heitorschueroff committed:
Update on "Fixed einsum compatibility issues"


TODO

- [x] add support for ellipsis broadcasting
- [x] fix corner case issues with sumproduct_pair
- [ ] update docs and add more comments
- [ ] add tests for error cases

[ghstack-poisoned]

heitorschueroff committed:
Update on "Fixed einsum compatibility/performance issues"


fixes #45854, #37628, #30194, #15671, #41467, #32591

TODO

- [x] add support for ellipsis broadcasting
- [x] fix corner case issues with sumproduct_pair
- [ ] update docs and add more comments
- [ ] add tests for error cases

[ghstack-poisoned]

heitorschueroff committed:
Update on "Fixed einsum compatibility/performance issues"


fixes #45854, #37628, #30194, #15671, #41467, #32591

TODO

- [x] add support for ellipsis broadcasting
- [x] fix corner case issues with sumproduct_pair
- [x] update docs and add more comments
- [ ] add tests for error cases

[ghstack-poisoned]

heitorschueroff committed:
Update on "Fixed einsum compatibility/performance issues"


This PR makes torch.einsum compatible with numpy.einsum except for the sublist input option NumPy has requested here https://github.com/pytorch/pytorch/issues/21412. It also fixed 2 performance issues linked below and adds a check for reducing to torch.dot instead of torch.bmm which is faster in some cases.

fixes #45854, #37628, #30194, #15671, #41467, #32591

TODO

- [x] add support for ellipsis broadcasting
- [x] fix corner case issues with sumproduct_pair
- [x] update docs and add more comments
- [x] add tests for error cases

[ghstack-poisoned]

mruberry commented on aten/src/ATen/native/Linear.cpp:
```diff
@@ -136,241 +136,319 @@ static Tensor sumproduct_pair(const Tensor& left_, const Tensor& right_, IntArra
+// 1. Parse equation to extract the labels for each input operand and output
+// 2. Unsqueeze missing dimensions from input operands and permute to align them
+// 3. Compute result by multiplying input operands and summing contraction
+//    dimensions We do the last part by reducing to bmm, there is room for
```
Maybe file and cite a follow-up performance issue for an investigation or TODO into the possible performance improvements here?

mruberry commented on aten/src/ATen/native/Linear.cpp:
```diff
@@ -136,241 +136,319 @@ static Tensor sumproduct_pair(const Tensor& left_, const Tensor& right_, IntArra
+  constexpr int ELLIPSIS = '.';
+
+  // Find arrow (->) to split equation into lhs and rhs
+  std::size_t arrow_pos = equation.find("->");
```
const auto

mruberry commented on aten/src/ATen/native/Linear.cpp:
```diff
@@ -136,241 +136,319 @@ static Tensor sumproduct_pair(const Tensor& left_, const Tensor& right_, IntArra
+  std::vector<std::vector<int>> op_labels(operands.size());
+  bool found_ell = false;
+  std::size_t curr_op = 0;
+  for (std::size_t i = 0; i < lhs.length(); ++i) {
```
Technically `i` would be `std::string::size_type`. You can get around the type mismatch by writing

```
const auto lhs_length = lhs.length()
for (auto i = decltype(lhs_length){0}; i < lhs_length; ++i) { 
```

If you really wanted to be fancy you could even use functions like `std::result_of` to avoid explicitly creating `lhs_length`, but I think it makes the code more readable.

mruberry commented on aten/src/ATen/native/Linear.cpp:
```diff
@@ -136,241 +136,319 @@ static Tensor sumproduct_pair(const Tensor& left_, const Tensor& right_, IntArra
+      case ',':
+        // Move onto next operand
+        TORCH_CHECK(
+            ++curr_op < operands.size(),
```
Separate the increment to curr_op from the TORCH_CHECK to make this more readable.

mruberry commented on test/test_linalg.py:
```diff
@@ -921,6 +921,143 @@ def test_nuclear_norm_exceptions_old(self, device):
+        def check(equation, *operands):
+            ref = np.einsum(equation, *[operand.cpu().numpy() for operand in operands])
+            res = torch.einsum(equation, operands)
+            self.assertEqual(res.shape, ref.shape)
```
I think this check is redundant with the tensors being equal below?

mruberry commented on torch/functional.py:
```diff
@@ -265,17 +265,19 @@ def einsum(equation, *operands):
-           Ellipses `...` represent a fixed number of dimensions. If the right hand side is inferred,
-           the ellipsis dimensions are at the beginning of the output.
+    equation (string): The subscripts for the Einstein summation specified as comma separated list of
+            subscript labels (lower case letters [a, z]). The left-hand-side must specify the subscripts for each
```
"left-hand_side" of the equation

mruberry commented on torch/functional.py:
```diff
@@ -265,17 +265,19 @@ def einsum(equation, *operands):
-           the ellipsis dimensions are at the beginning of the output.
+    equation (string): The subscripts for the Einstein summation specified as comma separated list of
+            subscript labels (lower case letters [a, z]). The left-hand-side must specify the subscripts for each
+            dimension of the input operands, separating operands by comma (,). The right-hand-side follows after an
```
"with each operand separated by a comma (","). "

mruberry replied:
"right-hand-side" of the equation, starting after an arrow ("->"), specifies

mruberry commented on torch/functional.py:
```diff
@@ -265,17 +265,19 @@ def einsum(equation, *operands):
+    equation (string): The subscripts for the Einstein summation specified as comma separated list of
+            subscript labels (lower case letters [a, z]). The left-hand-side must specify the subscripts for each
+            dimension of the input operands, separating operands by comma (,). The right-hand-side follows after an
+            arrow (->) and specifies the subscript labels for the output. These must appear in at least one operand.
```
"These must appear in at least one operand." -> "At least one operand must appear."

On the lhs, rhs, or both?

mruberry commented on torch/functional.py:
```diff
@@ -265,17 +265,19 @@ def einsum(equation, *operands):
+            subscript labels (lower case letters [a, z]). The left-hand-side must specify the subscripts for each
+            dimension of the input operands, separating operands by comma (,). The right-hand-side follows after an
+            arrow (->) and specifies the subscript labels for the output. These must appear in at least one operand.
+            If the arrow (->) and right-hand-side are ommited, the output is implicitly defined as the alphabetically
```
The arrow ("->") may be omitted, in which case the output is

mruberry commented on torch/functional.py:
```diff
@@ -265,17 +265,19 @@ def einsum(equation, *operands):
+            dimension of the input operands, separating operands by comma (,). The right-hand-side follows after an
+            arrow (->) and specifies the subscript labels for the output. These must appear in at least one operand.
+            If the arrow (->) and right-hand-side are ommited, the output is implicitly defined as the alphabetically
+            sorted list of all subscript labels appearing exactly once in the left-hand-side. The subscript labels
```
Provide a brief parenthetical like: (i.e. "ijk" is equivalent to ...)

mruberry commented on torch/functional.py:
```diff
@@ -265,17 +265,19 @@ def einsum(equation, *operands):
-           If an index appears several times for the same operand, a diagonal is taken.
-           Ellipses `...` represent a fixed number of dimensions. If the right hand side is inferred,
-           the ellipsis dimensions are at the beginning of the output.
+    equation (string): The subscripts for the Einstein summation specified as comma separated list of
```
It's wild how much documentation was packed into this argument description, and how little is in the function description itself. Maybe we can move some of this there and improve the description with examples?

heitorschueroff committed:
Update on "Fixed einsum compatibility/performance issues"


This PR makes torch.einsum compatible with numpy.einsum except for the sublist input option NumPy has requested here https://github.com/pytorch/pytorch/issues/21412. It also fixed 2 performance issues linked below and adds a check for reducing to torch.dot instead of torch.bmm which is faster in some cases.

fixes #45854, #37628, #30194, #15671

fixes #41467 with benchmark below
```python
import torch
from torch.utils.benchmark import Timer

a = torch.randn(10000, 100, 101, device='cuda')
b = torch.randn(10000, 101, 3, device='cuda')

c = torch.randn(10000, 100, 1, device='cuda')
d = torch.randn(10000, 100, 1, 3, device='cuda')

print(Timer(
    stmt='torch.einsum("bij,bjf->bif", a, b)',
    globals={'a': a, 'b': b}
).blocked_autorange())

print()

print(Timer(
    stmt='torch.einsum("bic,bicf->bif", c, d)',
    globals={'c': c, 'd': d}
).blocked_autorange())
```
```
<torch.utils.benchmark.utils.common.Measurement object at 0x7fa37c413850>
torch.einsum("bij,bjf->bif", a, b)
  Median: 4.53 ms
  IQR:    0.00 ms (4.53 to 4.53)
  45 measurements, 1 runs per measurement, 1 thread

<torch.utils.benchmark.utils.common.Measurement object at 0x7fa37c413700>
torch.einsum("bic,bicf->bif", c, d)
  Median: 63.86 us
  IQR:    1.52 us (63.22 to 64.73)
  4 measurements, 1000 runs per measurement, 1 thread
```

fixes #32591 with benchmark below
```python
import torch
from torch.utils.benchmark import Timer

a = torch.rand(1, 1, 16, 2, 16, 2, 16, 2, 2, 2, 2, device="cuda")
b = torch.rand(729, 1, 1, 2, 1, 2, 1, 2, 2, 2, 2, device="cuda")

print(Timer(
    stmt='(a * b).sum(dim = (-3, -2, -1))',
    globals={'a': a, 'b': b}
).blocked_autorange())

print()

print(Timer(
    stmt='torch.einsum("...ijk, ...ijk -> ...", a, b)',
    globals={'a': a, 'b': b}
).blocked_autorange())
```
```
<torch.utils.benchmark.utils.common.Measurement object at 0x7efe0de28850>
(a * b).sum(dim = (-3, -2, -1))
  Median: 17.86 ms
  2 measurements, 10 runs per measurement, 1 thread

<torch.utils.benchmark.utils.common.Measurement object at 0x7efe0de286a0>
torch.einsum("...ijk, ...ijk -> ...", a, b)
  Median: 296.11 us
  IQR:    1.38 us (295.42 to 296.81)
  662 measurements, 1 runs per measurement, 1 thread
```

TODO

- [x] add support for ellipsis broadcasting
- [x] fix corner case issues with sumproduct_pair
- [ ] update docs and add more comments
- [x] add tests for error cases

[ghstack-poisoned]

heitorschueroff commented on aten/src/ATen/native/Linear.cpp:
```diff
@@ -136,241 +136,321 @@ static Tensor sumproduct_pair(const Tensor& left_, const Tensor& right_, IntArra
+
+    for (int label : labels) {
+      if (label == ELLIPSIS) {
+        has_ellipsis = true;
```
Add comment explaining the +1 and why we need the ell_num_dim.

heitorschueroff committed:
Update on "Fixed einsum compatibility/performance issues"


This PR makes torch.einsum compatible with numpy.einsum except for the sublist input option NumPy has requested here https://github.com/pytorch/pytorch/issues/21412. It also fixed 2 performance issues linked below and adds a check for reducing to torch.dot instead of torch.bmm which is faster in some cases.

fixes #45854, #37628, #30194, #15671

fixes #41467 with benchmark below
```python
import torch
from torch.utils.benchmark import Timer

a = torch.randn(10000, 100, 101, device='cuda')
b = torch.randn(10000, 101, 3, device='cuda')

c = torch.randn(10000, 100, 1, device='cuda')
d = torch.randn(10000, 100, 1, 3, device='cuda')

print(Timer(
    stmt='torch.einsum("bij,bjf->bif", a, b)',
    globals={'a': a, 'b': b}
).blocked_autorange())

print()

print(Timer(
    stmt='torch.einsum("bic,bicf->bif", c, d)',
    globals={'c': c, 'd': d}
).blocked_autorange())
```
```
<torch.utils.benchmark.utils.common.Measurement object at 0x7fa37c413850>
torch.einsum("bij,bjf->bif", a, b)
  Median: 4.53 ms
  IQR:    0.00 ms (4.53 to 4.53)
  45 measurements, 1 runs per measurement, 1 thread

<torch.utils.benchmark.utils.common.Measurement object at 0x7fa37c413700>
torch.einsum("bic,bicf->bif", c, d)
  Median: 63.86 us
  IQR:    1.52 us (63.22 to 64.73)
  4 measurements, 1000 runs per measurement, 1 thread
```

fixes #32591 with benchmark below
```python
import torch
from torch.utils.benchmark import Timer

a = torch.rand(1, 1, 16, 2, 16, 2, 16, 2, 2, 2, 2, device="cuda")
b = torch.rand(729, 1, 1, 2, 1, 2, 1, 2, 2, 2, 2, device="cuda")

print(Timer(
    stmt='(a * b).sum(dim = (-3, -2, -1))',
    globals={'a': a, 'b': b}
).blocked_autorange())

print()

print(Timer(
    stmt='torch.einsum("...ijk, ...ijk -> ...", a, b)',
    globals={'a': a, 'b': b}
).blocked_autorange())
```
```
<torch.utils.benchmark.utils.common.Measurement object at 0x7efe0de28850>
(a * b).sum(dim = (-3, -2, -1))
  Median: 17.86 ms
  2 measurements, 10 runs per measurement, 1 thread

<torch.utils.benchmark.utils.common.Measurement object at 0x7efe0de286a0>
torch.einsum("...ijk, ...ijk -> ...", a, b)
  Median: 296.11 us
  IQR:    1.38 us (295.42 to 296.81)
  662 measurements, 1 runs per measurement, 1 thread
```

TODO

- [x] add support for ellipsis broadcasting
- [x] fix corner case issues with sumproduct_pair
- [ ] update docs and add more comments
- [x] add tests for error cases

[ghstack-poisoned]

heitorschueroff committed:
Update on "Fixed einsum compatibility/performance issues"


This PR makes torch.einsum compatible with numpy.einsum except for the sublist input option NumPy has requested here https://github.com/pytorch/pytorch/issues/21412. It also fixed 2 performance issues linked below and adds a check for reducing to torch.dot instead of torch.bmm which is faster in some cases.

fixes #45854, #37628, #30194, #15671

fixes #41467 with benchmark below
```python
import torch
from torch.utils.benchmark import Timer

a = torch.randn(10000, 100, 101, device='cuda')
b = torch.randn(10000, 101, 3, device='cuda')

c = torch.randn(10000, 100, 1, device='cuda')
d = torch.randn(10000, 100, 1, 3, device='cuda')

print(Timer(
    stmt='torch.einsum("bij,bjf->bif", a, b)',
    globals={'a': a, 'b': b}
).blocked_autorange())

print()

print(Timer(
    stmt='torch.einsum("bic,bicf->bif", c, d)',
    globals={'c': c, 'd': d}
).blocked_autorange())
```
```
<torch.utils.benchmark.utils.common.Measurement object at 0x7fa37c413850>
torch.einsum("bij,bjf->bif", a, b)
  Median: 4.53 ms
  IQR:    0.00 ms (4.53 to 4.53)
  45 measurements, 1 runs per measurement, 1 thread

<torch.utils.benchmark.utils.common.Measurement object at 0x7fa37c413700>
torch.einsum("bic,bicf->bif", c, d)
  Median: 63.86 us
  IQR:    1.52 us (63.22 to 64.73)
  4 measurements, 1000 runs per measurement, 1 thread
```

fixes #32591 with benchmark below
```python
import torch
from torch.utils.benchmark import Timer

a = torch.rand(1, 1, 16, 2, 16, 2, 16, 2, 2, 2, 2, device="cuda")
b = torch.rand(729, 1, 1, 2, 1, 2, 1, 2, 2, 2, 2, device="cuda")

print(Timer(
    stmt='(a * b).sum(dim = (-3, -2, -1))',
    globals={'a': a, 'b': b}
).blocked_autorange())

print()

print(Timer(
    stmt='torch.einsum("...ijk, ...ijk -> ...", a, b)',
    globals={'a': a, 'b': b}
).blocked_autorange())
```
```
<torch.utils.benchmark.utils.common.Measurement object at 0x7efe0de28850>
(a * b).sum(dim = (-3, -2, -1))
  Median: 17.86 ms
  2 measurements, 10 runs per measurement, 1 thread

<torch.utils.benchmark.utils.common.Measurement object at 0x7efe0de286a0>
torch.einsum("...ijk, ...ijk -> ...", a, b)
  Median: 296.11 us
  IQR:    1.38 us (295.42 to 296.81)
  662 measurements, 1 runs per measurement, 1 thread
```

TODO

- [x] add support for ellipsis broadcasting
- [x] fix corner case issues with sumproduct_pair
- [ ] update docs and add more comments
- [x] add tests for error cases

[ghstack-poisoned]

mruberry commented on torch/functional.py:
```diff
@@ -262,76 +262,100 @@ def lu_unpack(LU_data, LU_pivots, unpack_data=True, unpack_pivots=True):
-    torch.Size([2, 3, 5, 4])
-"""
+    Computes multilinear expressions (i.e. sums of products) specified using the Einstein summation convention.
+
```
An example here would probably be helpful.

mruberry commented on torch/functional.py:
```diff
@@ -262,76 +262,100 @@ def lu_unpack(LU_data, LU_pivots, unpack_data=True, unpack_pivots=True):
-    >>> torch.einsum('...ij->...ji', A).shape # batch permute
-    torch.Size([2, 3, 5, 4])
-"""
+    Computes multilinear expressions (i.e. sums of products) specified using the Einstein summation convention.
```
I'm not sure what a "multilinear expression" is. Also, are these equations really written using "Einstein notation?" My understand of Einstein notation isn't perfect but that description seems like a stretch. The [Wikipedia article](https://en.wikipedia.org/wiki/Einstein_notation) doesn't seem to have a 1:1 match with this equation, either. What do you think?

heitorschueroff replied:
You're right, It's based on Einstein notation but not the same and to be honest I'm not entirely sure what a multilinear expression is either, I guess it's a system of linear equations with multiple lines? haha. Just an idea but, maybe we should get someone more familiar with the math behind each operator to review docs besides just the engineer.

mruberry replied:
Maybe, but the typical reader won't know these terms, either. The documentation should break down what the operation does in terms that any reader can understand (possibly after consulting Wikipedia).

mruberry commented on torch/functional.py:
```diff
@@ -262,76 +262,100 @@ def lu_unpack(LU_data, LU_pivots, unpack_data=True, unpack_pivots=True):
+
+    The Einstein summation convention provides a succint way of specifying many linear algebra operations by
+    labeling each dimension of the input operands and specifying which dimensions are to be summed over.
+    Einsum provides two modes: implicit mode computes the classical Einstein summation while explicit mode
```
"two modes" makes it sound like there's a `mode` parameter.

mruberry commented on torch/functional.py:
```diff
@@ -262,76 +262,100 @@ def lu_unpack(LU_data, LU_pivots, unpack_data=True, unpack_pivots=True):
+
+        The :attr:`equation` string specifies the subscripts (lower case letters `['a', 'z']`) for the Eintein summation.
+
+        In implicit mode, :attr:`equation` specifies the subscripts for each dimension of the input :attr:`operands` in
```
What's actually computed still isn't clear at this point. Is there a mathematical formula that can be written before this? Then the notation could relate to that formula.

mruberry commented on torch/functional.py:
```diff
@@ -262,76 +262,100 @@ def lu_unpack(LU_data, LU_pivots, unpack_data=True, unpack_pivots=True):
+        dimensions under the subscript for the operand must match in size and the diagonal of the corresponding dimensions
+        will be taken. The dimensions that appear exactly once in the :attr:`equation` will be part of the output Tensor,
+        sorted in increasing alphabetical order. The dimensions that are not part of the output (appear more than once)
+        will be summed over.
```
What does it mean to "sum over" a dimension or dimensions?

mruberry commented on torch/functional.py:
```diff
@@ -262,76 +262,100 @@ def lu_unpack(LU_data, LU_pivots, unpack_data=True, unpack_pivots=True):
+        order, separating subcripts for each operand by a comma (','), e.g. `'ij,jk'` specify subscripts for two 2D
+        operands. The dimensions labeled with the same subscript must be broadcastable, that is, their size must either
+        match or be `1`. The exception is if a subscript is repeated for the same input operand, in which case the
+        dimensions under the subscript for the operand must match in size and the diagonal of the corresponding dimensions
```
I'm not sure what "dimensions under the subscript" means

mruberry commented on torch/functional.py:
```diff
@@ -262,76 +262,100 @@ def lu_unpack(LU_data, LU_pivots, unpack_data=True, unpack_pivots=True):
+        sorted in increasing alphabetical order. The dimensions that are not part of the output (appear more than once)
+        will be summed over.
+
+        In explicit mode, the same rules for the input operands in implicit mode apply but the output dimensions can be
```
Is it just that the dimensions can be reordered in the output? How should a user understand the "default" ordering/naming of output dimensions?

JackCaoG commented:
Hi @heitorschueroff , xla side fix is ready. I will merge it once this change merged to pytorch.

heitorschueroff committed:
Update on "Fixed einsum compatibility/performance issues"


This PR makes torch.einsum compatible with numpy.einsum except for the sublist input option as requested here https://github.com/pytorch/pytorch/issues/21412. It also fixed 2 performance issues linked below and adds a check for reducing to torch.dot instead of torch.bmm which is faster in some cases.

fixes #45854, #37628, #30194, #15671

fixes #41467 with benchmark below
```python
import torch
from torch.utils.benchmark import Timer

a = torch.randn(10000, 100, 101, device='cuda')
b = torch.randn(10000, 101, 3, device='cuda')

c = torch.randn(10000, 100, 1, device='cuda')
d = torch.randn(10000, 100, 1, 3, device='cuda')

print(Timer(
    stmt='torch.einsum("bij,bjf->bif", a, b)',
    globals={'a': a, 'b': b}
).blocked_autorange())

print()

print(Timer(
    stmt='torch.einsum("bic,bicf->bif", c, d)',
    globals={'c': c, 'd': d}
).blocked_autorange())
```
```
<torch.utils.benchmark.utils.common.Measurement object at 0x7fa37c413850>
torch.einsum("bij,bjf->bif", a, b)
  Median: 4.53 ms
  IQR:    0.00 ms (4.53 to 4.53)
  45 measurements, 1 runs per measurement, 1 thread

<torch.utils.benchmark.utils.common.Measurement object at 0x7fa37c413700>
torch.einsum("bic,bicf->bif", c, d)
  Median: 63.86 us
  IQR:    1.52 us (63.22 to 64.73)
  4 measurements, 1000 runs per measurement, 1 thread
```

fixes #32591 with benchmark below
```python
import torch
from torch.utils.benchmark import Timer

a = torch.rand(1, 1, 16, 2, 16, 2, 16, 2, 2, 2, 2, device="cuda")
b = torch.rand(729, 1, 1, 2, 1, 2, 1, 2, 2, 2, 2, device="cuda")

print(Timer(
    stmt='(a * b).sum(dim = (-3, -2, -1))',
    globals={'a': a, 'b': b}
).blocked_autorange())

print()

print(Timer(
    stmt='torch.einsum("...ijk, ...ijk -> ...", a, b)',
    globals={'a': a, 'b': b}
).blocked_autorange())
```
```
<torch.utils.benchmark.utils.common.Measurement object at 0x7efe0de28850>
(a * b).sum(dim = (-3, -2, -1))
  Median: 17.86 ms
  2 measurements, 10 runs per measurement, 1 thread

<torch.utils.benchmark.utils.common.Measurement object at 0x7efe0de286a0>
torch.einsum("...ijk, ...ijk -> ...", a, b)
  Median: 296.11 us
  IQR:    1.38 us (295.42 to 296.81)
  662 measurements, 1 runs per measurement, 1 thread
```

TODO

- [x] add support for ellipsis broadcasting
- [x] fix corner case issues with sumproduct_pair
- [x] update docs and add more comments
- [x] add tests for error cases

[ghstack-poisoned]

heitorschueroff committed:
Update on "Fixed einsum compatibility/performance issues"


This PR makes torch.einsum compatible with numpy.einsum except for the sublist input option as requested here https://github.com/pytorch/pytorch/issues/21412. It also fixed 2 performance issues linked below and adds a check for reducing to torch.dot instead of torch.bmm which is faster in some cases.

fixes #45854, #37628, #30194, #15671

fixes #41467 with benchmark below
```python
import torch
from torch.utils.benchmark import Timer

a = torch.randn(10000, 100, 101, device='cuda')
b = torch.randn(10000, 101, 3, device='cuda')

c = torch.randn(10000, 100, 1, device='cuda')
d = torch.randn(10000, 100, 1, 3, device='cuda')

print(Timer(
    stmt='torch.einsum("bij,bjf->bif", a, b)',
    globals={'a': a, 'b': b}
).blocked_autorange())

print()

print(Timer(
    stmt='torch.einsum("bic,bicf->bif", c, d)',
    globals={'c': c, 'd': d}
).blocked_autorange())
```
```
<torch.utils.benchmark.utils.common.Measurement object at 0x7fa37c413850>
torch.einsum("bij,bjf->bif", a, b)
  Median: 4.53 ms
  IQR:    0.00 ms (4.53 to 4.53)
  45 measurements, 1 runs per measurement, 1 thread

<torch.utils.benchmark.utils.common.Measurement object at 0x7fa37c413700>
torch.einsum("bic,bicf->bif", c, d)
  Median: 63.86 us
  IQR:    1.52 us (63.22 to 64.73)
  4 measurements, 1000 runs per measurement, 1 thread
```

fixes #32591 with benchmark below
```python
import torch
from torch.utils.benchmark import Timer

a = torch.rand(1, 1, 16, 2, 16, 2, 16, 2, 2, 2, 2, device="cuda")
b = torch.rand(729, 1, 1, 2, 1, 2, 1, 2, 2, 2, 2, device="cuda")

print(Timer(
    stmt='(a * b).sum(dim = (-3, -2, -1))',
    globals={'a': a, 'b': b}
).blocked_autorange())

print()

print(Timer(
    stmt='torch.einsum("...ijk, ...ijk -> ...", a, b)',
    globals={'a': a, 'b': b}
).blocked_autorange())
```
```
<torch.utils.benchmark.utils.common.Measurement object at 0x7efe0de28850>
(a * b).sum(dim = (-3, -2, -1))
  Median: 17.86 ms
  2 measurements, 10 runs per measurement, 1 thread

<torch.utils.benchmark.utils.common.Measurement object at 0x7efe0de286a0>
torch.einsum("...ijk, ...ijk -> ...", a, b)
  Median: 296.11 us
  IQR:    1.38 us (295.42 to 296.81)
  662 measurements, 1 runs per measurement, 1 thread
```

TODO

- [x] add support for ellipsis broadcasting
- [x] fix corner case issues with sumproduct_pair
- [x] update docs and add more comments
- [x] add tests for error cases

[ghstack-poisoned]

heitorschueroff committed:
Update on "Fixed einsum compatibility/performance issues"


This PR makes torch.einsum compatible with numpy.einsum except for the sublist input option as requested here https://github.com/pytorch/pytorch/issues/21412. It also fixed 2 performance issues linked below and adds a check for reducing to torch.dot instead of torch.bmm which is faster in some cases.

fixes #45854, #37628, #30194, #15671

fixes #41467 with benchmark below
```python
import torch
from torch.utils.benchmark import Timer

a = torch.randn(10000, 100, 101, device='cuda')
b = torch.randn(10000, 101, 3, device='cuda')

c = torch.randn(10000, 100, 1, device='cuda')
d = torch.randn(10000, 100, 1, 3, device='cuda')

print(Timer(
    stmt='torch.einsum("bij,bjf->bif", a, b)',
    globals={'a': a, 'b': b}
).blocked_autorange())

print()

print(Timer(
    stmt='torch.einsum("bic,bicf->bif", c, d)',
    globals={'c': c, 'd': d}
).blocked_autorange())
```
```
<torch.utils.benchmark.utils.common.Measurement object at 0x7fa37c413850>
torch.einsum("bij,bjf->bif", a, b)
  Median: 4.53 ms
  IQR:    0.00 ms (4.53 to 4.53)
  45 measurements, 1 runs per measurement, 1 thread

<torch.utils.benchmark.utils.common.Measurement object at 0x7fa37c413700>
torch.einsum("bic,bicf->bif", c, d)
  Median: 63.86 us
  IQR:    1.52 us (63.22 to 64.73)
  4 measurements, 1000 runs per measurement, 1 thread
```

fixes #32591 with benchmark below
```python
import torch
from torch.utils.benchmark import Timer

a = torch.rand(1, 1, 16, 2, 16, 2, 16, 2, 2, 2, 2, device="cuda")
b = torch.rand(729, 1, 1, 2, 1, 2, 1, 2, 2, 2, 2, device="cuda")

print(Timer(
    stmt='(a * b).sum(dim = (-3, -2, -1))',
    globals={'a': a, 'b': b}
).blocked_autorange())

print()

print(Timer(
    stmt='torch.einsum("...ijk, ...ijk -> ...", a, b)',
    globals={'a': a, 'b': b}
).blocked_autorange())
```
```
<torch.utils.benchmark.utils.common.Measurement object at 0x7efe0de28850>
(a * b).sum(dim = (-3, -2, -1))
  Median: 17.86 ms
  2 measurements, 10 runs per measurement, 1 thread

<torch.utils.benchmark.utils.common.Measurement object at 0x7efe0de286a0>
torch.einsum("...ijk, ...ijk -> ...", a, b)
  Median: 296.11 us
  IQR:    1.38 us (295.42 to 296.81)
  662 measurements, 1 runs per measurement, 1 thread
```

TODO

- [x] add support for ellipsis broadcasting
- [x] fix corner case issues with sumproduct_pair
- [x] update docs and add more comments
- [x] add tests for error cases

[ghstack-poisoned]

heitorschueroff committed:
Update on "Fixed einsum compatibility/performance issues"


This PR makes torch.einsum compatible with numpy.einsum except for the sublist input option as requested here https://github.com/pytorch/pytorch/issues/21412. It also fixed 2 performance issues linked below and adds a check for reducing to torch.dot instead of torch.bmm which is faster in some cases.

fixes #45854, #37628, #30194, #15671

fixes #41467 with benchmark below
```python
import torch
from torch.utils.benchmark import Timer

a = torch.randn(10000, 100, 101, device='cuda')
b = torch.randn(10000, 101, 3, device='cuda')

c = torch.randn(10000, 100, 1, device='cuda')
d = torch.randn(10000, 100, 1, 3, device='cuda')

print(Timer(
    stmt='torch.einsum("bij,bjf->bif", a, b)',
    globals={'a': a, 'b': b}
).blocked_autorange())

print()

print(Timer(
    stmt='torch.einsum("bic,bicf->bif", c, d)',
    globals={'c': c, 'd': d}
).blocked_autorange())
```
```
<torch.utils.benchmark.utils.common.Measurement object at 0x7fa37c413850>
torch.einsum("bij,bjf->bif", a, b)
  Median: 4.53 ms
  IQR:    0.00 ms (4.53 to 4.53)
  45 measurements, 1 runs per measurement, 1 thread

<torch.utils.benchmark.utils.common.Measurement object at 0x7fa37c413700>
torch.einsum("bic,bicf->bif", c, d)
  Median: 63.86 us
  IQR:    1.52 us (63.22 to 64.73)
  4 measurements, 1000 runs per measurement, 1 thread
```

fixes #32591 with benchmark below
```python
import torch
from torch.utils.benchmark import Timer

a = torch.rand(1, 1, 16, 2, 16, 2, 16, 2, 2, 2, 2, device="cuda")
b = torch.rand(729, 1, 1, 2, 1, 2, 1, 2, 2, 2, 2, device="cuda")

print(Timer(
    stmt='(a * b).sum(dim = (-3, -2, -1))',
    globals={'a': a, 'b': b}
).blocked_autorange())

print()

print(Timer(
    stmt='torch.einsum("...ijk, ...ijk -> ...", a, b)',
    globals={'a': a, 'b': b}
).blocked_autorange())
```
```
<torch.utils.benchmark.utils.common.Measurement object at 0x7efe0de28850>
(a * b).sum(dim = (-3, -2, -1))
  Median: 17.86 ms
  2 measurements, 10 runs per measurement, 1 thread

<torch.utils.benchmark.utils.common.Measurement object at 0x7efe0de286a0>
torch.einsum("...ijk, ...ijk -> ...", a, b)
  Median: 296.11 us
  IQR:    1.38 us (295.42 to 296.81)
  662 measurements, 1 runs per measurement, 1 thread
```

TODO

- [x] add support for ellipsis broadcasting
- [x] fix corner case issues with sumproduct_pair
- [x] update docs and add more comments
- [x] add tests for error cases

[ghstack-poisoned]

mruberry commented on torch/functional.py:
```diff
@@ -262,76 +262,102 @@ def lu_unpack(LU_data, LU_pivots, unpack_data=True, unpack_pivots=True):
-    torch.Size([2, 3, 5, 4])
-"""
+    Sums the product of the elements of the input :attr:`operands` along dimensions specified using a notation
+    based on Einstein summation convention.
```
"based on the Einstein summation convention"

mruberry commented on torch/functional.py:
```diff
@@ -262,76 +262,102 @@ def lu_unpack(LU_data, LU_pivots, unpack_data=True, unpack_pivots=True):
+    Sums the product of the elements of the input :attr:`operands` along dimensions specified using a notation
+    based on Einstein summation convention.
+
+    Einsum allows computing many common multi-dimensional, linear algebraic array operations by representing them
```
no comma needed here

mruberry commented on torch/functional.py:
```diff
@@ -262,76 +262,102 @@ def lu_unpack(LU_data, LU_pivots, unpack_data=True, unpack_pivots=True):
+    based on Einstein summation convention.
+
+    Einsum allows computing many common multi-dimensional, linear algebraic array operations by representing them
+    in a short-hand format based on the Eintein summation, given by :attr:`equation`. The details of this format are
```
"the Einstein summation convention" to be consistent with the above

mruberry commented on torch/functional.py:
```diff
@@ -262,76 +262,102 @@ def lu_unpack(LU_data, LU_pivots, unpack_data=True, unpack_pivots=True):
+
+    Einsum allows computing many common multi-dimensional, linear algebraic array operations by representing them
+    in a short-hand format based on the Eintein summation, given by :attr:`equation`. The details of this format are
+    described in the section below, but the general idea is to label every dimension of the input :attr:`operands`
```
"described below"

mruberry reviewed (approved):
Docs look a lot better and the benchmarks look good. Made a few small grammar comments.

heitorschueroff committed:
Update on "Fixed einsum compatibility/performance issues"


This PR makes torch.einsum compatible with numpy.einsum except for the sublist input option as requested here https://github.com/pytorch/pytorch/issues/21412. It also fixed 2 performance issues linked below and adds a check for reducing to torch.dot instead of torch.bmm which is faster in some cases.

fixes #45854, #37628, #30194, #15671

fixes #41467 with benchmark below
```python
import torch
from torch.utils.benchmark import Timer

a = torch.randn(10000, 100, 101, device='cuda')
b = torch.randn(10000, 101, 3, device='cuda')

c = torch.randn(10000, 100, 1, device='cuda')
d = torch.randn(10000, 100, 1, 3, device='cuda')

print(Timer(
    stmt='torch.einsum("bij,bjf->bif", a, b)',
    globals={'a': a, 'b': b}
).blocked_autorange())

print()

print(Timer(
    stmt='torch.einsum("bic,bicf->bif", c, d)',
    globals={'c': c, 'd': d}
).blocked_autorange())
```
```
<torch.utils.benchmark.utils.common.Measurement object at 0x7fa37c413850>
torch.einsum("bij,bjf->bif", a, b)
  Median: 4.53 ms
  IQR:    0.00 ms (4.53 to 4.53)
  45 measurements, 1 runs per measurement, 1 thread

<torch.utils.benchmark.utils.common.Measurement object at 0x7fa37c413700>
torch.einsum("bic,bicf->bif", c, d)
  Median: 63.86 us
  IQR:    1.52 us (63.22 to 64.73)
  4 measurements, 1000 runs per measurement, 1 thread
```

fixes #32591 with benchmark below
```python
import torch
from torch.utils.benchmark import Timer

a = torch.rand(1, 1, 16, 2, 16, 2, 16, 2, 2, 2, 2, device="cuda")
b = torch.rand(729, 1, 1, 2, 1, 2, 1, 2, 2, 2, 2, device="cuda")

print(Timer(
    stmt='(a * b).sum(dim = (-3, -2, -1))',
    globals={'a': a, 'b': b}
).blocked_autorange())

print()

print(Timer(
    stmt='torch.einsum("...ijk, ...ijk -> ...", a, b)',
    globals={'a': a, 'b': b}
).blocked_autorange())
```
```
<torch.utils.benchmark.utils.common.Measurement object at 0x7efe0de28850>
(a * b).sum(dim = (-3, -2, -1))
  Median: 17.86 ms
  2 measurements, 10 runs per measurement, 1 thread

<torch.utils.benchmark.utils.common.Measurement object at 0x7efe0de286a0>
torch.einsum("...ijk, ...ijk -> ...", a, b)
  Median: 296.11 us
  IQR:    1.38 us (295.42 to 296.81)
  662 measurements, 1 runs per measurement, 1 thread
```

TODO

- [x] add support for ellipsis broadcasting
- [x] fix corner case issues with sumproduct_pair
- [x] update docs and add more comments
- [x] add tests for error cases

[ghstack-poisoned]

Darktex commented on aten/src/ATen/native/Linear.cpp:
```diff
@@ -136,241 +136,331 @@ static Tensor sumproduct_pair(const Tensor& left_, const Tensor& right_, IntArra
       }
     }
+
+    TORCH_CHECK(
```
I think this check is breaking Opacus

mruberry replied:
Hey @Darktex, could you elaborate?

heitorschueroff replied:
It could be, I didn't notice this at first but the previous version does not enforce this rule so this check is BC breaking. @mruberry I can send a quick patch to remove this rule and allow ellipsis to be reduced or I can add a BC breaking note. (numpy has this enforcement but it doesn't seem necessary).

heitorschueroff committed:
Update on "Fixed einsum compatibility/performance issues"


This PR makes torch.einsum compatible with numpy.einsum except for the sublist input option as requested here https://github.com/pytorch/pytorch/issues/21412. It also fixed 2 performance issues linked below and adds a check for reducing to torch.dot instead of torch.bmm which is faster in some cases.

### Update: BC Breaking changes
This PR introduced a backward compatibility breaking change. Before, dimensions covered by an ellipsis could be contracted but after this PR, if ellipsis are given for the input operands, they must also be part of the output tensor and cannot be contracted (summed out).

fixes #45854, #37628, #30194, #15671

fixes #41467 with benchmark below
```python
import torch
from torch.utils.benchmark import Timer

a = torch.randn(10000, 100, 101, device='cuda')
b = torch.randn(10000, 101, 3, device='cuda')

c = torch.randn(10000, 100, 1, device='cuda')
d = torch.randn(10000, 100, 1, 3, device='cuda')

print(Timer(
    stmt='torch.einsum("bij,bjf->bif", a, b)',
    globals={'a': a, 'b': b}
).blocked_autorange())

print()

print(Timer(
    stmt='torch.einsum("bic,bicf->bif", c, d)',
    globals={'c': c, 'd': d}
).blocked_autorange())
```
```
<torch.utils.benchmark.utils.common.Measurement object at 0x7fa37c413850>
torch.einsum("bij,bjf->bif", a, b)
  Median: 4.53 ms
  IQR:    0.00 ms (4.53 to 4.53)
  45 measurements, 1 runs per measurement, 1 thread

<torch.utils.benchmark.utils.common.Measurement object at 0x7fa37c413700>
torch.einsum("bic,bicf->bif", c, d)
  Median: 63.86 us
  IQR:    1.52 us (63.22 to 64.73)
  4 measurements, 1000 runs per measurement, 1 thread
```

fixes #32591 with benchmark below
```python
import torch
from torch.utils.benchmark import Timer

a = torch.rand(1, 1, 16, 2, 16, 2, 16, 2, 2, 2, 2, device="cuda")
b = torch.rand(729, 1, 1, 2, 1, 2, 1, 2, 2, 2, 2, device="cuda")

print(Timer(
    stmt='(a * b).sum(dim = (-3, -2, -1))',
    globals={'a': a, 'b': b}
).blocked_autorange())

print()

print(Timer(
    stmt='torch.einsum("...ijk, ...ijk -> ...", a, b)',
    globals={'a': a, 'b': b}
).blocked_autorange())
```
```
<torch.utils.benchmark.utils.common.Measurement object at 0x7efe0de28850>
(a * b).sum(dim = (-3, -2, -1))
  Median: 17.86 ms
  2 measurements, 10 runs per measurement, 1 thread

<torch.utils.benchmark.utils.common.Measurement object at 0x7efe0de286a0>
torch.einsum("...ijk, ...ijk -> ...", a, b)
  Median: 296.11 us
  IQR:    1.38 us (295.42 to 296.81)
  662 measurements, 1 runs per measurement, 1 thread
```

TODO

- [x] add support for ellipsis broadcasting
- [x] fix corner case issues with sumproduct_pair
- [x] update docs and add more comments
- [x] add tests for error cases

Differential Revision: [D24860367](https://our.internmc.facebook.com/intern/diff/D24860367)

[ghstack-poisoned]

heitorschueroff commented:
Closing because of ghstack issue
